src/API.jl

```diff
@@ -1099,703 +1098,16 @@ function precompile(ctx::Context, pkgs::Vector{PackageSpec}; internal_call::Bool
-
-    print_lock = ctx.io isa Base.LibuvStream ? ctx.io.lock::ReentrantLock : ReentrantLock()
-    first_started = Base.Event()
-    printloop_should_exit::Bool = !fancyprint # exit print loop immediately if not fancy printing
-    interrupted_or_done = Base.Event()
-
-    ansi_moveup(n::Int) = string("\e[", n, "A")
-    ansi_movecol1 = "\e[1G"
-    ansi_cleartoend = "\e[0J"
-    ansi_cleartoendofline = "\e[0K"
-    ansi_enablecursor = "\e[?25h"
-    ansi_disablecursor = "\e[?25l"
-    n_done::Int = 0
-    n_already_precomp::Int = 0
-    n_loaded::Int = 0
-    interrupted = false
-
-    function handle_interrupt(err, in_printloop = false)
-        notify(interrupted_or_done)
-        in_printloop || wait(t_print) # wait to let the print loop cease first
-        if err isa InterruptException
-            lock(print_lock) do
-                println(io, " Interrupted: Exiting precompilation...")
-            end
-            interrupted = true
-            return true
-        else
-            return false
-        end
-    end
-
-    std_outputs = Dict{Base.PkgId,String}()
-    taskwaiting = Set{Base.PkgId}()
-    pkgspidlocked = Dict{Base.PkgId,String}()
-    pkg_liveprinted = nothing
-
-    function monitor_std(pkg, pipe; single_requested_pkg=false)
-        try
-            liveprinting = false
-            while !eof(pipe)
-                str = readline(pipe, keep=true)
-                if single_requested_pkg && (liveprinting || !isempty(str))
-                    lock(print_lock) do
-                        if !liveprinting
-                            printpkgstyle(io, :Info, "Given $(pkg.name) was explicitly requested, output will be shown live $ansi_cleartoendofline",
-                                color = Base.info_color())
-                            liveprinting = true
-                            pkg_liveprinted = pkg
-                        end
-                        print(io, ansi_cleartoendofline, str)
-                    end
-                end
-                std_outputs[pkg] = string(get(std_outputs, pkg, ""), str)
-                if !in(pkg, taskwaiting) && occursin("waiting for IO to finish", str)
-                    !fancyprint && lock(print_lock) do
-                        println(io, pkg.name, color_string(" Waiting for background task / IO / timer.", Base.warn_color()))
-                    end
-                    push!(taskwaiting, pkg)
-                end
-                if !fancyprint && in(pkg, taskwaiting)
-                    lock(print_lock) do
-                        print(io, str)
-                    end
-                end
-            end
-        catch err
-            err isa InterruptException || rethrow()
-        end
+    activate(dirname(ctx.env.project_file)) do
+        pkgs_name = String[pkg.name for pkg in pkgs]
+        return Base.PrecompilePkgs.precompilepkgs(pkgs_name; internal_call, strict, warn_loaded, timing, _from_loading, flags_cacheflags, io=ctx.io)
     end
```

Comment:
Why is the activate needed? Pkg.precompile calls `instantiate` before this so if this is needed it needs to happen earlier?

Reply:
I think this was an issue before because the Base loading code needs to have the packages getting precompiled in the load path to look stuff up but if you create a Context with some random environment path, that won't be the case (unless we activate it here)